test/delegations.spec.ts

```diff
@@ -0,0 +1,156 @@
+import 'mocha';
+
+import * as _ from "lodash";
+import Web3 from "web3";
+declare const web3: Web3;
+
+import BN from "bn.js";
+import {
+    defaultDriverOptions,
+    BANNING_LOCK_TIMEOUT,
+    Driver,
+    expectRejected,
+    Participant
+} from "./driver";
+import chai from "chai";
+chai.use(require('chai-bn')(BN));
+chai.use(require('./matchers'));
+
+const expect = chai.expect;
+const assert = chai.assert;
+
+import {bn, evmIncreaseTime} from "./helpers";
+import {ETHEREUM_URL, Web3Driver} from "../eth";
+import {
+    committeeChangedEvents,
+    delegatedEvents,
+    stakedEvents,
+    delegatedStakeChangedEvents
+} from "./event-parsing";
+
+const baseStake = 100;
+
+describe('delegations-contract', async () => {
+
+
+    it('should only accept stake notifications from the staking contract', async () => {
+        const d = await Driver.new();
+
+        const rogueStakingContract = await d.newStakingContract(d.delegations.address, d.erc20.address);
+
+        const participant = d.newParticipant();
+
+        await expectRejected(participant.stake(5, rogueStakingContract), "should not accept notifications from an address other than the staking contract");
+        await participant.stake(5);
+        await d.contractRegistry.set("staking", rogueStakingContract.address);
+        await participant.stake(5, rogueStakingContract)
+
+        // TODO - to check stakeChangeBatch use a mock staking contract that would satisfy the interface but would allow sending stakeChangeBatch when there are no rewards to distribue
+        // await expectRejected(d.delegations.stakeChangeBatch([d.accounts[0]], [1], [true], [1], {from: nonStakingAddr}), "should not accept notifications from an address other than the staking contract");
+        // await d.delegations.stakeChangeBatch([d.accounts[0]], [1], [true], [1], {from: stakingAddr});
+    });
+
+    it('emits DelegatedStakeChanged and Delegated on delegation changes', async () => {
+        const d = await Driver.new();
+
+        const p1 = d.newParticipant();
+        let r = await p1.stake(100);
+        expect(r).to.have.a.delegatedStakeChangedEvent({
+            addr: p1.address,
+            selfStake: bn(100),
+            delegatedStake: bn(100),
+            delegators: [p1.address],
+            delegatorTotalStakes: [bn(100)]
+        });
+
+        const p2 = d.newParticipant();
+        r = await p2.delegate(p1);
```

Comment:
This should also emit `delegatedStakeChangedEvent` right? `p2` has zero stake, so maybe it doesn't have to emit.

Reply:
right - it will not fire the event. added to test.